src/Voron/Data/CompactTrees/CompactTree.cs

```diff
@@ -1897,52 +1448,40 @@ private void FindPageFor(EncodedKey key, ref IteratorCursorState cstate)
 
             key.ChangeDictionary(state.Header->DictionaryId);
 
-            FindPageFor(ref cstate, ref state, key);
-        }
-
-        private void FindPageFor(ReadOnlySpan<byte> key, ref IteratorCursorState cstate, EncodedKey encodedKey)
-        {
-            cstate._pos = -1;
-            cstate._len = 0;
-            PushPage(_state.RootPage, ref cstate);
-
-            ref var state = ref cstate._stk[cstate._pos];
-
-            encodedKey.Set(key);
-            encodedKey.ChangeDictionary(state.Header->DictionaryId);
-
-            FindPageFor(ref cstate, ref state, encodedKey);
+            FindPageFor(ref cstate, ref state, key, tryRecompress);
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void FindPageFor(ref IteratorCursorState cstate, ref CursorState state, EncodedKey encodedKey)
+        private void FindPageFor(ref IteratorCursorState cstate, ref CursorState state, CompactKey compactKey, bool tryRecompress = false)
         {
             while (state.Header->PageFlags.HasFlag(CompactPageFlags.Branch))
             {
-                SearchPageAndPushNext(encodedKey, ref cstate);
-                state = ref cstate._stk[cstate._pos];
-            }
-            SearchInCurrentPage(encodedKey, ref state);
-        }
+                // PERF: We aim to carry out the dictionary migration during the search process. This is because even though it may take longer,
+                // we can offset the cost of future key re-encoding along the way. We could choose to do this either before or after
+                // adding/searching, but performing it proactively guarantees that we avoid the need for re-encoding at this stage.
+                if (tryRecompress && _state.TreeDictionaryId != state.Header->DictionaryId)
+                    TryRecompressPage(ref state);
```

Comment:
This deals with all the branch pages, and will proactively move them
Given that we aren't deleting dictionaries, what is the current trigger for generating a new one?

Reply:
Also, you are updating the _branches_, but I think that the only time we'll recompress a leaf page is when we are out of space, no?

https://github.com/ravendb/ravendb/blob/c62101b8381fbe33c0825aa9997c76f8cb3b475e/src/Voron/Data/CompactTrees/CompactTree.cs#L958

If we are proactive, shouldn't we always do that when we modify a leaf page? (Note, we have to deal with the newly compressed page being _too big_ to fit the page in that scenario.

Reply:
Mhhh, it would make sense to do that proactively too for leaf, even though it is going to happen when splits happen (more often). 

When the recompress page fails, it just keep retrying on write until it is able to do so. We could add a marker for that, but it is not such a big deal; considering that the fallback is just reencode the key for that dictionary anyways... We are playing the game that eventually it will happen.

Reply:
The generation of new dictionaries follows a formula that makes it more difficult to trigger the bigger the tree is.